### ISSUE TITLE
Possible bug in ranges closer to "now" at m1

### ISSUE THREAD

**Issue by ivnnv** [Leo4815162342/dukascopy-node#109]
Hey, ive been trying to get `m1` candle data as closest as "now" as possible and it doesnt seems to work.
I have had a preliminar look at it could be a problem on `getClosestAvailableRange` range detection, but id need to dig deeper into it, you probably have a better idea on how is this happening and if it potentially has a bug.

![Screenshot 2023-01-24 at 17 57 14](https://user-images.githubusercontent.com/23552631/214439772-df50ebc7-cc2e-43ba-9466-ae73a7b8d206.png)

Trying to fetch 1m data from "~10 minutes ago" to "~1 minute ago" for example, it raises an error.

**Comment by Leo4815162342** [comment]
@ivnnv this potentially could happen because of the fact that `dukascopy` updates data on their feed by the end of every hour. 

As of the time for writing this message (~08:45 AM UTC), the `8 hour` data file is not there yet (only after 9:00 AM - the file will be create and available.):

https://datafeed.dukascopy.com/datafeed/BTCUSD/2023/00/26/08h_ticks.bi5 



<img width="952" alt="Screenshot 2023-01-26 at 09 40 12" src="https://user-images.githubusercontent.com/12486774/214792109-8a35bef4-4303-4fb9-a5e2-748f676ccbf1.png">

However the previous file is there

https://datafeed.dukascopy.com/datafeed/BTCUSD/2023/00/26/07h_ticks.bi5

**Comment by ivnnv** [comment]
oh, thats so unfortunate, so do you think  theres no "clean" way to request "last minute data" consistently?

My guess is only using something like https://github.com/ismailfer/dukascopy-api-websocket to keep a process asking for the last-minute data every minute.

**Comment by Leo4815162342** [comment]
@ivnnv sorry for a delayed response.

I think I found a way to get real-time data. I'm working on its experimental implementation - https://github.com/Leo4815162342/dukascopy-node/pull/111 - once it's merged and released, I'll ping you so you could test it

**Comment by ivnnv** [comment]
hey @Leo4815162342 thanks for the update and be working on the workaround already!
Feel free to release a beta version that I could test myself too if you want to, or yeah the final version if you are confident about it once its done 👍🏽

**Comment by Leo4815162342** [comment]
@ivnnv the new version has been released  - `1.30.1`

it now exports experimental `getCurrentRates` method which retrieves current/latest data from the source that powers the realtime charts.

```js
const { getCurrentRates } = require("dukascopy-node");

getCurrentRates({
  instrument: "btcusd",
  timeframe: "m1",
  limit: 10,
  format: "json",
  volumes: false,
}).then(console.log);
```

for now it's not hooked up to CLI and is not part of main `getHistoricalRates` method.

**Comment by ivnnv** [comment]
hey @Leo4815162342 I have tested this and it seems is working correctly, im getting what it seems a Typescript error tho:

![Screenshot 2023-02-01 at 00 50 57](https://user-images.githubusercontent.com/23552631/215962604-8bcb340b-cd93-45cd-8f7c-e0b35d42f319.png)
-
![Screenshot 2023-02-01 at 00 50 52](https://user-images.githubusercontent.com/23552631/215962592-769e9757-de1f-4464-8b59-7f219030ef9f.png)

its complaining about the `array` format but its actually working and retrieving the data in OHLCV format.

Then I think theres another issue with the `limit` parameter.
For example: If I pass from/to with dates 10 minutes away AND the limit is set to 30, it will return 30 items, but the from/to should take preference and return only 10 items (thats why im "overriding" the limit to "minutesUntilNow" in my screenshot above, so im making sure thats the diff between from and to)

Other than that, thank you for implementing this so quickly, appreciate it.

**Comment by Leo4815162342** [comment]
@ivnnv 

the new beta `1.31.0-beta.4` contains few updates:

`dates` now takes precedence over `limit` (if dates is provided, limit will be ignored).
`dates.to` is optional and defaults to now.

`dates.from` and `dates.to` can either be `Date`, `dateIsoString` or `unixTimestamp`

See some examples:

```typescript
import { getCurrentRates, CurrentRatesConfig } from "dukascopy-node";

// will fetch last 10 m1 candles (if no limit is provided it will default to 10)
const getCurrentRatesParamsWithLimit: CurrentRatesConfig = {
  instrument: "btcusd",
  timeframe: "m1",
  format: "array",
  limit: 10,
};

// will fetch last all available m1 candles from 2023-02-03
const getCurrentRatesParamsWithFromDate: CurrentRatesConfig = {
  instrument: "btcusd",
  timeframe: "m1",
  format: "json",
  dates: {
    from: new Date("2023-02-03"),
  },
};

// will fetch last all available m1 candles from 2023-02-03 until 2021-02-01
const getCurrentRatesParamsWithFromAndToDate: CurrentRatesConfig = {
  instrument: "btcusd",
  timeframe: "m30",
  format: "array",
  dates: {
    from: new Date("2021-01-30"),
    to: new Date("2021-02-01"),
  },
};

getCurrentRates(getCurrentRatesParamsWithLimit).then((rates) => {
  console.log("number of fetched candles", rates.length);
  console.log(rates);
});
```

**Comment by ivnnv** [comment]
Hi Leo, I just teste this and I think still some bugs on dates and types:

· Dates with only 36 minute diff returns 15.000 items, including pre-"from":
https://runkit.com/reg/63dd8d5222f75800084fdffc

· `CurrentRatesConfig` doesnt exists, and `CurrentRatesArgsArrayItem` exists but its not actually exported nor an `|` in `Config` type.
`error TS2459: Module '"dukascopy-node"' declares 'CurrentRatesArgsArrayItem' locally, but it is not exported.`

**Comment by Leo4815162342** [comment]
@ivnnv you need version ‘ 1.31.0-beta.4’

**Comment by ivnnv** [comment]
@Leo4815162342 oh sorry I thought the last update was published with the beta changes included.

Yeah I can sucessfully reproduce now and the types are ok, I can even get very "early currents" like one day before.

Speaking about this case, while doing some edge case testing, I found that some records are missing and its probably because `ignoreFlats` is not implemented yet right? TS also complains if try to add as on `getHistoricalRates`.

```
import dayjs from "dayjs";
import { getCurrentRates, CurrentRatesConfig } from "dukascopy-node";

const dates = {
  from: dayjs().subtract(1441, "minute").startOf("minute").toDate(),
  to: dayjs().subtract(1, "minute").startOf("minute").toDate(),
};

const getCurrentParams: CurrentRatesConfig = {
  dates,
  instrument: "btcusd", <-- if used `eurusc` no records are returned at all on weekends
  // ignoreFlats: false, <-- cant be implemented on currents?
  timeframe: "m1",
  format: "array",
};

const data = await getCurrentRates(getCurrentParams);
console.log(data.length); // 1219 instead 1440
```

**Comment by Leo4815162342** [comment]
@ivnnv yeah, the underlying real-time chart API does not handle flats in the same way as we're having for `getHistoricalRates`, so ATM it's not supported

**Comment by ivnnv** [comment]
@Leo4815162342 in that case I think this issue can be closed once you apply this changes onto master!

Other than that, could be `ignoreFlats` be custom implemented as `giving a repeated record (same values as the last retrieved one for each "gap") until a new one is found`? I could try to create a PR for that if you think that could be useful.

**Comment by Leo4815162342** [comment]
@ivnnv awesome, that sounds like a plausible idea - feel free to kick of a PR 🙌🏻 

the changes have been merged to master
